### PR TITLE
C++: Model varargs in IR, Part I

### PR DESCRIPTION
**Pull request by dbartol**
This change introduces a new synthesized `IRVariable` in every varargs function. This variable represents the entire set of arguments passed to the ellipsis by the caller. We give it an opaque type big enough hold all of the arguments passed by the largest vararg call in the database. It is treated just like any other parameter. It is initialized the same, it has indirect buffers, etc.

I had to introduce a couple new APIs to `Call` and `Function`. The QLDoc comments should explain these. I added tests for these new APIs as well.

The next step will be to change the IR generation for the `va_*` macros to manipulate the ellipsis parameter.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedFunction.qll [hunk 1]
```diff
@@ -16,6 +16,32 @@ private import TranslatedStmt
...
+            sum(int argIndex |
+              call.isEllipsisArgumentIndex(argIndex)
+            |
+              call.getArgument(argIndex).getType().getSize()
+            )
+        |
+          callSize
+        )
+      or
+      variableSize = 8
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedFunction.qll]: Is there a reason the minimum is 8 bytes and not, say, 1 byte?

Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/raw/internal/TranslatedFunction.qll]: I think I was assuming that we'd be adding alignment padding to any small values passed as varargs, the same way actual code generation would, so I padded the 1 byte out to the size of a pointer. In retrospect, the layout of the buffer probably won't ever be apparent in the IR, so this doesn't matter.

File: cpp/ql/src/semmle/code/cpp/exprs/Call.qll [hunk 2]
```diff
@@ -78,6 +85,32 @@ abstract class Call extends Expr, NameQualifiableElement {
...
+    else result = argIndex
+  }
+
+  /**
+   * Holds if the argument specified by `index` is an argument to the `...` of a varargs function.
+   */
+  final predicate isEllipsisArgumentIndex(int index) {
+    exists(getArgument(index)) and index >= getEllipsisParameterIndex()
+  }
+
```
[the remark this answers is not in the document]
Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/exprs/Call.qll]: Fair enough. I've moved them into a new `VarArgs.qll` file in the IR construction directory. I've left the tests I added in place.

File: cpp/ql/src/semmle/code/cpp/exprs/Call.qll [hunk 3]
```diff
@@ -2,6 +2,13 @@ import semmle.code.cpp.exprs.Expr
 import semmle.code.cpp.Function
 private import semmle.code.cpp.dataflow.EscapesTree
 
+/**
+ * Gets the index of the `...` parameter, if any.
+ */
+private int getEllipsisParameterIndex(RoutineType type) {
+  type.getParameterType(result) instanceof UnknownType
```
[the remark this answers is not in the document]
Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/exprs/Call.qll]: Done.

File: cpp/ql/src/semmle/code/cpp/exprs/Call.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/exprs/Call.qll]: Looks good!

File: cpp/ql/src/semmle/code/cpp/exprs/Call.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/exprs/Call.qll]: (It's now in `getEllipsisParameterIndexForRoutineType()`)

**Pull request merged**